File: leet-code/101.对称二叉树.js
/*
 * @lc app=leetcode.cn id=101 lang=javascript
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function (root) {
  if (!root) return true

  // 使用递归遍历左右子树 递归三部曲
  // 1. 确定递归的参数 root.left root.right和返回值true false 
  const compareNode = (left, right) => {
    // 2. 确定终止条件 空的情况
    if ((left === null && right !== null) || (left !== null && right === null)) {
      return false;
    }
    if (left === null && right === null) {
      return true;
    }
    if (left.val !== right.val) {
      return false;
    }
    // 3. 确定单层递归逻辑
    // 外侧: 左节点的左孩子<===>右节点的右孩子
    let outSide = compareNode(left.left, right.right);
    // 内侧: 左节点的右孩子<===>右节点的左孩子
    let inSide = compareNode(left.right, right.left);
    return outSide && inSide;
  };
  return compareNode(root.left, root.right)
}
// @lc code=end

File: leet-code-rs/src/solutions/101_对称二叉树.rs
/*
 * @lc app=leetcode.cn id=101 lang=rust
 *
 * [101] 对称二叉树
 */
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

#[allow(unused)]
impl Solution {
    pub fn is_match(
        left: &Option<Rc<RefCell<TreeNode>>>,
        right: &Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (left, right) {
            (Some(l), Some(r)) => {
                if l.borrow().val != r.borrow().val {
                    return false;
                }
                Self::is_match(&l.borrow().left, &r.borrow().right)
                    && Self::is_match(&l.borrow().right, &r.borrow().left)
            }
            (None, None) => true,
            _ => false,
        }
    }
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if root.is_none() {
            return true;
        }
        let root = root.unwrap();
        let left = &root.borrow().left;
        let right = &root.borrow().right;

        Self::is_match(left, right)
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_is_symmetric() {
        //assert_eq!(Solution::is_symmetric(vec![]),[]);
        assert!(true)
    }
}
